bin/cheribsdtest/cheribsdtest_vm.c

```diff
@@ -2309,6 +2309,159 @@ CHERIBSDTEST(revoke_merge_quarantined,
 	cheribsdtest_success();
 }
 #undef NRES
+
+/*
+ * A simple test to confirm that revocation of a capability in a COW mapping
+ * affects only the caller's mapping.
+ */
+CHERIBSDTEST(cheri_revoke_cow_mapping,
+    "verify that revocation of a COW page triggers a copy",
+    .ct_check_skip = skip_need_cheri_revoke)
+{
+	void **block, **cap1, **cap2;
+	void *shadow, *torev;
+	ssize_t n;
+	size_t blocksz;
+	pid_t child;
+	int pd[2], res;
+	char ch, st[2];
+
+	/*
+	 * Use three pages for our heap.  The last page will be revoked.  The
+	 * first two pages contain a pointer into the third page; the first
+	 * page will be unmapped before revocation, while the second will remain
+	 * mapped.  This difference exercises different code paths in the
+	 * revoker.
+	 */
+	blocksz = 3 * PAGE_SIZE;
+	block = mmap(NULL, blocksz, PROT_READ | PROT_WRITE, MAP_ANON, -1, 0);
+	CHERIBSDTEST_VERIFY(block != MAP_FAILED);
+
+	torev = cheri_setbounds(block + 2 * PAGE_SIZE / sizeof(void *),
+	    PAGE_SIZE);
+	cap1 = cheri_setbounds(&block[0], PAGE_SIZE);
+	cap2 = cheri_setbounds(&block[PAGE_SIZE / sizeof(void *)], PAGE_SIZE);
+	*cap1 = *cap2 = cheri_andperm(torev, ~CHERI_PERM_SW_VMEM);
+
+	child = fork();
+	if (child == -1)
+		cheribsdtest_failure_errx("Fork failed; errno=%d", errno);
+	if (child == 0) {
+		/*
+		 * Quarantine the third page.
+		 */
+		if (cheri_revoke_get_shadow(CHERI_REVOKE_SHADOW_NOVMEM,
+		    torev, &shadow) != 0)
+			_exit(1);
+		memset(shadow, 0xff, cheri_getlen(shadow));
+
+		/*
+		 * Make sure that the first page is unmapped in the current
+		 * pmap.  munmap() would have the same effect of course, but
+		 * would also throw away the contents of the page.
```

Comment:
Hmm, I guess I don't fully understand how munmap() would be different here.  Is that munmap() would alter the behavior of the vm_map_entry in the child in a way?  I guess maybe you want the revocation pass to scan both pages in the backing VM object, but the revocation pass limits itself to the subset of each VM object that is mapped by the vm_map_entry's?  So if you did munmap() then the revoker would only attempt to scan the first page and not the second?

Reply:
We have two pages, each containing a capability referencing the third page. I want the revoker to visit both pages and revoke both of those capabilities. I'm trying to ensure that the revoker's pmap has a PTE for the second page but not the first. munmap() would destroy the PTE, but it would also create a hole in the vm_map, causing the revoker not to visit that page.

Reply:
... and that is really just a re-statement of what you wrote, yes. :)

Reply:
Might be worth expanding the comment a bit further then as I think as it stands now saying you want the page unmapped doesn't quite map to that, or at least unmapped is ambiguous.  I think of unmapped primarily as "logically removed from the address space", i.e. munmap, but I think here you mean "removed from the pmap map (PTEs)".  Just something about the vm_map_entry still being the same so the revoker scans it all, but getting the PTE cleared is probably sufficient?

Reply:
I reworded it to avoid mention of unmapping the page.